components/common/src/templating/hooks.rs

```diff
@@ -440,54 +440,81 @@ pub struct HookOutput<'a> {
 
 impl<'a> HookOutput<'a> {
     fn new(stdout_log: &'a Path, stderr_log: &'a Path) -> Self {
-        HookOutput { stdout_log_file: stdout_log,
-                     stderr_log_file: stderr_log, }
+        Self { stdout_log_file: stdout_log,
+               stderr_log_file: stderr_log, }
+    }
+
+    pub fn stdout(&self) -> Result<BufReader<File>> {
+        Ok(BufReader::new(File::open(&self.stdout_log_file)?))
     }
 
-    pub fn stdout(&self) -> Option<BufReader<File>> {
-        match File::open(&self.stdout_log_file) {
-            Ok(f) => Some(BufReader::new(f)),
-            Err(_) => None,
+    pub fn stdout_str(&self) -> Option<String> {
+        let result = self.stdout_str_impl();
+        if let Err(e) = &result {
+            error!("Failed to read {:?}, {}", self.stdout_log_file, e);
         }
+        result.ok()
+    }
+
+    pub fn stderr(&self) -> Result<BufReader<File>> {
+        Ok(BufReader::new(File::open(&self.stderr_log_file)?))
     }
 
-    pub fn stderr(&self) -> Option<BufReader<File>> {
-        match File::open(&self.stderr_log_file) {
-            Ok(f) => Some(BufReader::new(f)),
-            Err(_) => None,
+    pub fn stderr_str(&self) -> Option<String> {
+        let result = self.stderr_str_impl();
+        if let Err(e) = &result {
+            error!("Failed to read {:?}, {}", self.stderr_log_file, e);
         }
+        result.ok()
     }
 
-    fn stream_output<H: Hook>(&mut self, service_group: &str, process: &mut Child) {
-        let mut stdout_log =
-            File::create(&self.stdout_log_file).expect("couldn't create log output file");
-        let mut stderr_log =
-            File::create(&self.stderr_log_file).expect("couldn't create log output file");
+    /// Try to write the stdout and stderr of a process to stdout and to the specified log files.
+    fn output_streams<H: Hook>(&mut self, service_group: &str, process: &mut Child) {
+        let preamble_str = Self::stream_preamble::<H>(service_group);
+        if let Some(stdout) = &mut process.stdout {
+            Self::output_stream(&preamble_str, stdout, &self.stdout_log_file);
+        }
+        if let Some(stderr) = &mut process.stderr {
+            Self::output_stream(&preamble_str, stderr, &self.stderr_log_file);
+        }
+    }
 
-        let preamble_str = self.stream_preamble::<H>(service_group);
-        if let Some(ref mut stdout) = process.stdout {
-            for line in BufReader::new(stdout).lines() {
-                if let Ok(ref l) = line {
+    /// Try to write a stream to stdout and to `path`  
+    fn output_stream<R: Read>(preamble_str: &str, reader: R, path: &Path) {
```

Comment:
This is much better, thanks!

I noticed you used `impl Read`. Do we prefer that even in argument position if the trait bounds are simple enough?

Reply:
I'm actually wondering if we could handle this generally... basically, have a "tee" function that takes a `Read` and a `Write`... pass in stdout / stderr as the `Read` and a Writer dumping to the file of your choice as the `Write`.

That would also make all this way more testable.

Reply:
> I noticed you used `impl Read`. Do we prefer that even in argument position if the bounds are simple enough?

I definitely do, though I don't think we have any sort of agreement as a team to prefer it. I just think it's simpler and more readable. Any arguments against it?

Reply:
> Any arguments against it?

No arguments against it. I am not a huge fan of having three ways to specify the trait bounds: inline with the generic, in a `where` clause, and with `impl`. I wish the rust language had some guidelines on when each method is prefered.

Reply:
There might be something in https://rust-lang-nursery.github.io/api-guidelines/about.html, but I haven't been able to find it with a quick search.

Since the parameter/return position `impl Trait` approach is the simplest, but doesn't work in all cases, I prefer to use that when it can be used. For the other two, between inline and using a `where` clause, I think it comes down to readability. I only use the inline version when the bound is very short and simple. As a guideline, I would suggest using inline if the signature can fit in one line, otherwise put it in a `where` clause.